Adaptive Joint Rule Components

obucklin opened the pull request:
This changes the `DirectRule` and `CategoryRule` GH components to adapt dynamically to the joint typ input. The joint type passes the names of the beams' roles to the rule components, which registers the new input parameters. This should simplify defining `main_beam` and `cross_beam` and also help with future joints such as mortise/tenon and dovetail.

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [`x` ] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ `x`] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [ `x`] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [ `x`] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas_timber.datastructures.Beam`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added necessary documentation (if appropriate)

obucklin commented:
[Adaptive_GH_Components_DEMO.zip](https://github.com/gramaziokohler/compas_timber/files/14299633/Adaptive_GH_Components_DEMO.zip)
sample file

obucklin commented:
This is somewhat experimental, so we should test it before pushing into a release. I did a fair amount of testing with various joints, and it seems stable, but I'm not sure about all the various GH events that could impact it...

obucklin commented:
OK so when opening a saved file with the components, old connections to the beam or category inputs are deleted. I will look into a solution. I'll let you know when it's working, or let me know if you have an idea.

chenkasirer commented:
Super cool feature! 
Can you please rebase/merge main and take care of  linting and formatting so that it's nicer to review? thanks!

obucklin commented:
Ok so I solved the issue where connections were severed when opening a saved file with the new components. This has the added benefit of not severing connections if the joint type uses the same `beam_names`. Included is a new demo GH file.
[Adaptive_GH_Components_DEMO_V2.zip](https://github.com/gramaziokohler/compas_timber/files/14312477/Adaptive_GH_Components_DEMO_V2.zip)

obucklin commented:
> Super cool feature! Can you please rebase/merge main and take care of linting and formatting so that it's nicer to review? thanks!
> 
> Also please stick to PEP8 convention where possible, even in GH components (e.g. snake_case for functions, variables and arguments).

ok done

chenkasirer commented on src/compas_timber/ghpython/components/CT_Joint_Rule_Category/code.py:
```diff
@@ -3,7 +3,64 @@
+            ghenv.Component.Params.Input[len(ghenv.Component.Params.Input) - 1]
+        )
+    ghenv.Component.Params.OnParametersChanged()
+    ghenv.Component.ExpireSolution(False)
```
There's a module called `ghcomponent_helpers.py` under `ghpython, maybe move these functions there? that way you can reuse them in the other components instead of redefining

chenkasirer commented on src/compas_timber/ghpython/components/CT_Joint_Rule_Category/code.py:
```diff
@@ -3,7 +3,64 @@
+
+class DirectJointRule(component):
+    def RunScript(self, joint_options, *args):
+        if not joint_options:  # if no joint_options is input
+            clear_params()
+            return
+
+        register_params = False
+        if len(ghenv.Component.Params.Input) == len(joint_options.beam_names) + 1:
+            for i, name in enumerate(joint_options.beam_names):
+                if ghenv.Component.Params.Input[i + 1].Name != name:
+                    register_params = True
+                    break
+        else:
+            register_params = True
+        if register_params:  # if joint_options changes
+            if len(joint_options.beam_names) != 2:
+                self.AddRuntimeMessage(Error, "Component currently only supports joint types with 2 beams.")
+            clear_params()
+            for name in joint_options.beam_names:
+                add_param(name, "Input")
+
+        if len(ghenv.Component.Params.Input) != len(joint_options.beam_names) + 1 or len(args) != len(
+            joint_options.beam_names
+        ):  # something went wrong and the number of input parameters is wrong
+            self.AddRuntimeMessage(Warning, "Input parameter error.")
+            return
```
maybe this logic can be written more generically e.g. function that takes a list of param names (and whatever else it needs) and takes care of everything necessary to add them properly? then it could also be moved to the above said module and reused in the other components.

obucklin replied:
this has been moved to ghcomponent_helpers.py in method `manage_dynamic_params`

chenkasirer commented on src/compas_timber/ghpython/components/CT_Joint_Rule_Direct/code.py:
```diff
@@ -1,37 +1,97 @@
+            ghenv.Component.Params.Input[len(ghenv.Component.Params.Input) - 1]
+        )
+    ghenv.Component.Params.OnParametersChanged()
+    ghenv.Component.ExpireSolution(True)
```
see above

chenkasirer commented on src/compas_timber/ghpython/components/CT_Joint_Rule_Direct/code.py:
```diff
@@ -1,37 +1,97 @@
+            for i, name in enumerate(joint_options.beam_names):
+                if ghenv.Component.Params.Input[i + 1].Name != name:
+                    register_params = True
+                    break
+        else:
+            register_params = True
+        if register_params:  # if JointOptions changes
+            if len(joint_options.beam_names) != 2:
+                self.AddRuntimeMessage(Error, "Component currently only supports joint types with 2 beams.")
+            clear_params()
+            for name in joint_options.beam_names:
+                add_param(name, "Input")
+
+        if (
+            len(ghenv.Component.Params.Input) != len(joint_options.beam_names) + 1
+        ):  # something went wrong and the number of input parameters is wrong
+            self.AddRuntimeMessage(Warning, "Input parameter error.")
             return
-        Rules = []
-        for main, secondary in zip(MainBeam, SecondaryBeam):
-            topology, _, _ = ConnectionSolver().find_topology(main, secondary)
-            if topology != JointOptions.type.SUPPORTED_TOPOLOGY:
+
+        beams = []
+        create_rule = True
+        for i in range(len(ghenv.Component.Params.Input) - 1):
+            if not args[i]:
+                self.AddRuntimeMessage(
+                    Warning, "Input parameter {} failed to collect data.".format(joint_options.beam_names[i])
+                )
+                create_rule = False
+            else:
+                arg_beams = args[i]
+                if not isinstance(arg_beams, list):
+                    arg_beams = [arg_beams]
+                beams.append(arg_beams)
```
see above

chenkasirer commented on src/compas_timber/ghpython/components/CT_Joint_Options_FrenchRidgeLap/code.py:
```diff
@@ -10,6 +10,6 @@ def RunScript(self, Cutoff):
         if Cutoff:
             args["cutoff"] = Cutoff
-        options = JointOptions(FrenchRidgeLapJoint, **args)
+        options = JointOptions(FrenchRidgeLapJoint, ["top_beam", "bottom_beam"] ** args)
```
```suggestion
        options = JointOptions(FrenchRidgeLapJoint, ["top_beam", "bottom_beam"], **args)
```

chenkasirer commented on src/compas_timber/ghpython/components/CT_Joint_Options_LButt/code.py:
```diff
@@ -15,6 +15,6 @@ def RunScript(self, small_beam_butts, modify_cross, reject_i):
             args["reject_i"] = reject_i
 
-        options = JointOptions(LButtJoint, **args)
+        options = JointOptions(LButtJoint, ["main_beam", "cross_beam"], **args)
```
Sorry to nitpick but since these names end up on the components should we not use CamelCase here as we usually do?

obucklin replied:
done

chenkasirer commented on src/compas_timber/ghpython/components/CT_Joint_Options_LButt/code.py:
```diff
@@ -15,6 +15,6 @@ def RunScript(self, small_beam_butts, modify_cross, reject_i):
             args["reject_i"] = reject_i
 
-        options = JointOptions(LButtJoint, **args)
+        options = JointOptions(LButtJoint, ["main_beam", "cross_beam"], **args)
```
don't we need to add an argument to `JointOptions.__init__()` so that it accepts now the least of beam names?

obucklin replied:
somehow that got deleted in a merge. Now it's back!

chenkasirer commented on src/compas_timber/ghpython/components/CT_Joint_Rule_Category/code.py:
```diff
@@ -3,7 +3,64 @@
+            return
+
+        register_params = False
+        if len(ghenv.Component.Params.Input) == len(joint_options.beam_names) + 1:
```
I don't see you added `beam_names` attribute to `JointOptions`..?

obucklin commented:
@chenkasirer @jonashaldemann Hey this is ready for review. There are videos and an up-to-date example file.

chenkasirer commented:
> @chenkasirer @jonashaldemann Hey this is ready for review. There are videos and an up-to-date example file.

please make sure all actions finish successfully

chenkasirer commented:
generally this needs to be discussed and agreed upon between @obucklin and @jonashaldemann (and potentially others that are invested in the GH flow).

* Maybe find a better place for the videos.
* There's still some code style/naming issues here and there, try to stick to PEP8 where possible. 
* Test on MAC..

obucklin commented:
I went through the code for all the components and updated the formatting. 
I tested the adaptive functionality on mac and it works fine.
removed the videos, may upload to youtube.

Not sure about the actions...

chenkasirer reviewed (approved):
Pending:
1. @jonashaldemann is happy with this flow
2. another round of testing cause of the latest changes

LGTM.
